df_engine_extensions/src/dist_sql_query/mod.rs

```diff
@@ -53,8 +52,127 @@ pub trait ExecutableScanBuilder: fmt::Debug + Send + Sync + 'static {
+    /// `read_parallelism`.
+    pub read_parallelism: usize,
+
+    /// The schema and projection for read, the output data should match this
+    /// schema.
+    pub projected_schema: ProjectedSchema,
+
+    /// Predicate of the query.
+    pub predicate: PredicateRef,
+}
+
+impl TableScanContext {
+    pub fn new(
+        batch_size: usize,
+        read_parallelism: usize,
+        projected_schema: ProjectedSchema,
+        predicate: PredicateRef,
+    ) -> Self {
+        Self {
+            batch_size,
+            read_parallelism,
+            projected_schema,
+            predicate,
+        }
+    }
+}
+
+impl TryFrom<TableScanContext> for ceresdbproto::remote_engine::TableScanContext {
+    type Error = datafusion::error::DataFusionError;
+
+    fn try_from(value: TableScanContext) -> DfResult<Self> {
+        let pb_projected_schema = value
+            .projected_schema
+            .try_into()
+            .box_err()
+            .map_err(DataFusionError::External)?;
+
+        let pb_predicate = value
+            .predicate
+            .as_ref()
+            .try_into()
+            .box_err()
+            .map_err(DataFusionError::External)?;
+
+        Ok(Self {
+            batch_size: value.batch_size as u64,
+            read_parallelism: value.read_parallelism as u64,
+            projected_schema: Some(pb_projected_schema),
+            predicate: Some(pb_predicate),
+        })
+    }
+}
+
+impl TryFrom<ceresdbproto::remote_engine::TableScanContext> for TableScanContext {
+    type Error = datafusion::error::DataFusionError;
+
+    fn try_from(value: ceresdbproto::remote_engine::TableScanContext) -> DfResult<Self> {
+        let projected_schema = value
+            .projected_schema
+            .ok_or(DataFusionError::Internal(
+                "projected schema not found".to_string(),
+            ))?
+            .try_into()
+            .map_err(|e| {
+                DataFusionError::Internal(format!("failed to decode projected schema, err:{}", e))
+            })?;
+
+        let predicate = value
+            .predicate
+            .ok_or(DataFusionError::Internal("predicate not found".to_string()))?
+            .try_into()
+            .map_err(|e| {
+                DataFusionError::Internal(format!("failed to decode predicate, err:{}", e))
+            })?;
+
+        Ok(Self {
+            batch_size: value.batch_size as usize,
+            read_parallelism: value.read_parallelism as usize,
+            projected_schema,
+            predicate: Arc::new(predicate),
+        })
+    }
+}
+
+impl fmt::Debug for TableScanContext {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let exprs = self
+            .predicate
+            .exprs()
+            .iter()
+            .map(|expr| expr.to_string())
+            .collect::<Vec<_>>()
+            .join(",");
+        let predicate = format!("[{}]", exprs);
```

Comment:
`Predicate` seems derived from `Debug`, so shall we just use it directly?

Reply:
It is not so readable use default `Debug`...
I impl a custom debug dispaly for it?

Reply:
Ok. How about use `f.debug_list`, which receives an iterator, to avoid the heap memory allocation?

Reply:
Looks great!